### ISSUE TITLE
Please add support for `--stdin` argument

### ISSUE THREAD

**Issue by mgzenitech** [dotenv-linter/dotenv-linter#803]
Backstory: https://github.com/dotenv-linter/dotenv-linter/discussions/764#discussioncomment-10047505

For VSCode support we can use any-lint (https://github.com/tamayika/vscode-any-lint) extension with this config:
```json
"any-lint.linters": [
    {
      "args": [
        "-q",
        "${file}"
      ],
      "binPath": "dotenv-linter",
      "condition": "$.fileExtname.endsWith('.env')",
      "cwd": "${workspaceFolder}",
      "diagnostic": {
        "format": "${file}:${startLine} ${message}",
        "output": "stdout"
      },
      "name": "dotenv-linter",
      "on": [
        "change",
        "save"
      ]
    }
  ]
```

The only issue here is that `dotenv-linter` should support `--stdin` to be able to work with `onChange` event type and in general to be able to pass body of the file. More info in any-lint issue https://github.com/tamayika/vscode-any-lint/issues/1

**Comment by ccoVeille** [comment]
I don't get the need, passing a stdin + filename, so here you are somehow asking for another way to pass a file.

Why not, maybe, but then ...

Why doesn't open a ticket on anylint to pass file as argument.

I'm sure I'm missing something, so I'm not saying there is no need, but I'm asking why the problem shouldn't be solved on any-lint side

**Comment by mgzenitech** [comment]
> I don't get the need, passing a stdin + filename, so here you are somehow asking for another way to pass a file.
> 
> Why not, maybe, but then ...
> 
> Why doesn't open a ticket on anylint to pass file as argument.
> 
> I'm sure I'm missing something, so I'm not saying there is no need, but I'm asking why the problem shouldn't be solved on any-lint side

Maybe because VSCode doesn't store in-edit file that's not saved yet? Not sure, I'm not extension developer. This change would allow integration with VSCode

**Comment by ccoVeille** [comment]
> Maybe because VSCode doesn't store in-edit file that's not saved yet? Not sure, I'm not extension developer. This change would allow better integration with VSCode.

Indeed, maybe @tamayika could explain.

**Comment by tamayika** [comment]
Yes, unsaved content is not saved to file, so linter can not read it from file.

**Comment by ccoVeille** [comment]
Thanks for confirming it, it makes sense now.

**Comment by DDtKey** [comment]
Hi there 👋 

Actually, I think it makes a lot of sense to support `stdin`. 
This way, pipes can also be used. E.g:

```sh
some_command | dotenv-linter --stdin
```

**Comment by JonathanLorimer** [comment]
This would be extremely helpful. Especially given workflows that involve sops (which doesn't properly quote dotenv output, which breaks multiline variables)

**Comment by mgzenitech** [comment]
yeah, once someone adds this I'll update config for any-lint too.

Right now only onsave event supported: https://github.com/tamayika/vscode-any-lint/blob/main/examples/dotenv-linter.jsonc

**Comment by vivek378521** [comment]
I'm new here, I'd like to work on this - I went through the PR: https://github.com/dotenv-linter/dotenv-linter/pull/772/files to understand how we can add new arguments, but I see we also want to store the filename?

**Comment by jorgehermo9** [comment]
nit: The idiomatic way of accepting stdin instead of a file in a cli should be using `-` as a filename.

Example: `dotenv-linter -`

**Comment by DDtKey** [comment]
> Will allow that just dotenv-linter (with no file arg) defaults to stdin (which is how cat, for example, works)

Currently the default behavior is something like `dotenv-linter .` and it scans all env files within the current directory.

It's more classic behavior for linters (cargo fmt, clippy and many others).

**Comment by mgzenitech** [comment]
> > The idiomatic way of accepting stdin instead of a file in a cli should be using - as a filename.
> 
> That's definitely a well established pattern. I don't mind supporting this. However we can even allow both: either `-` (as special filename) or `--stdin` (explicit option)
> 
> > Will allow that just dotenv-linter (with no file arg) defaults to stdin (which is how cat, for example, works)
> 
> This would be breaking change at least. Currently the default behavior is something like `dotenv-linter .` and it scans all env files within the directory.
> 
> It's quite common default behavior for linters (e.g cargo fmt, clippy and many others).

Will you also accept file name argument when doing stdin? Because otherwise you will not have enough information to display in error messages. And I think without file path information next to error messages any-lint will fail

**Comment by DDtKey** [comment]
> Will you also accept file name argument when doing stdin? Because otherwise you will not have enough information to display in error messages. And I think without file path information next to error messages any-lint will fail

I'm not sure I got your concerns correctly. Could you elaborate? 
I don't see any reason to require a file name since `stdin input may not be associated with any file. Also, it expects the contents of 1 env file, so I don't see where the ambiguous output could be coming from

**Comment by mgzenitech** [comment]
> > Will you also accept file name argument when doing stdin? Because otherwise you will not have enough information to display in error messages. And I think without file path information next to error messages any-lint will fail
> 
> I'm not sure I got your concerns correctly. Could you elaborate? I don't see any reason to require a file name since `stdin` input may not be associated with any file. Also, it expects the contents of 1 env file, so I don't see where the ambiguous output could be coming from

So right now dotenv-linter shows error messages in this format:
```
${file}:${startLine} ${ruleId}: ${message}
```
if we pipe the content without known file what will be in `${file}`?

**Comment by charan-kumar-137** [comment]
> if we pipe the content without known file what will be in `${file}`?

It will be blank. STDIN is itself a file and it doesn't need any name associated to it.

**Comment by SirSam21** [comment]
> if we pipe the content without known file what will be in `${file}`?
> 
> Any-lint needs error messages to contain path to the file even when passed via stdin to inform VScode correctly. More info @ [tamayika/vscode-any-lint#1](https://github.com/tamayika/vscode-any-lint/issues/1)

Support could be added for --stdin-filename but is there a reason to do so? Any-lint should already have access to the filename through vscode.textDocument from which the data is passed to stdin. Wouldn't the best solution be to have any-lint default to that file name when using onChange?

**Comment by mgzenitech** [comment]
> > if we pipe the content without known file what will be in `${file}`?
> > Any-lint needs error messages to contain path to the file even when passed via stdin to inform VScode correctly. More info @ [tamayika/vscode-any-lint#1](https://github.com/tamayika/vscode-any-lint/issues/1)
> 
> Support could be added for --stdin-filename but is there a reason to do so? Any-lint should already have access to the filename through vscode.textDocument from which the data is passed to stdin. Wouldn't the best solution be to have any-lint default to that file name when using onChange?

@tamayika would it be possible to update any-lint with these changes ^^^ Or at least explain the reasoning why --stdin-filename is required :)

**Comment by tamayika** [comment]
Yes, it is possible to make `filename` placeholder optional in `format` setting.

**Comment by mgzenitech** [comment]
> Yes, it is possible to make `filename` placeholder optional in `format` setting.

cool, so if @tamayika can patch the extension to get the filename automatically then this only needs `--stdin` argument only. Any ETA when these things can be implemented?

**Comment by tamayika** [comment]
I have supported it in `0.4.1`.
Also, I added `extraArgs` to support `change` and `save` event in one configuration.
Please see [eslint example](https://github.com/tamayika/vscode-any-lint/blob/main/examples/eslint.jsonc)

**Comment by mgzenitech** [comment]
> @mgzenitech I have supported it in `0.4.1`. Also, I added `extraArgs` to support `change` and `save` event in one configuration. Please see [eslint example](https://github.com/tamayika/vscode-any-lint/blob/main/examples/eslint.jsonc)

How to configure it to use different formats for save and change?

this is linter config:

```json
      ...
      "args": [
        "-q"
      ],
      "binPath": "dotenv-linter",
      "extraArgs": {
        "change": [
          "--stdin"
        ],
        "save": [
          "${file}"
        ]
      },
      "on": [
        "save",
        "change"
      ],
      ...
```

and this is diagnostic format:
```json
"format": "${file}:${startLine} ${ruleId}: ${message}",
```

for stdint I think dotenv will provide something like "${startLine} ${ruleId}: ${message}"

**Comment by tamayika** [comment]
`${file}` matches to 0 or more characters, so empty string can be matched.

**Comment by Cherepanov-Pavel** [comment]
Hi guys, any updates here?
